Feat/sentry environment

PatrickRoethlisberger opened the pull request:
**Adds [Sentry Environments](https://docs.sentry.io/product/sentry-basics/environments/) support**

**Changes proposed in this pull request:**
This PR adds support for setting a custom sentry environment while unifying the default to the base URL.

**Screenshot**
Setting in the admin interface with default:
![Admin Setting](https://user-images.githubusercontent.com/9606082/181256051-a282299f-1eb1-4e51-99e0-75bf390638e0.png)

Environment set in request to sentry instance:
![Default Env in Request](https://user-images.githubusercontent.com/9606082/181256482-68b2bf2f-dbfc-42ca-b8ce-4ff55aa48845.png)

Settings in admin interface with custom environment:
![Admin Setting](https://user-images.githubusercontent.com/9606082/181257525-b9aad441-6ed3-48a3-8586-1f66e2da6f28.png)

Custom Environment in request to sentry instance:
![Custom Env in Request](https://user-images.githubusercontent.com/9606082/181257729-6e051266-7542-4918-a507-63041489287a.png)


**Confirmed**

- [x] Frontend changes: tested on a local Flarum installation.
- [X] Backend changes: tested on a local Flarum installation.

PatrickRoethlisberger committed:
Add admin setting for the sentry environment

PatrickRoethlisberger committed:
Add default for fof-sentry.environment setting

PatrickRoethlisberger committed:
Add environment set in settings to sentry calls

dsevillamartin commented on extend.php:
```diff
@@ -58,5 +59,6 @@
     (new Flarum\Settings())
-        ->default('fof-sentry.monitor_performance', 0),
+        ->default('fof-sentry.monitor_performance', 0)
+        ->default('fof-sentry.environment', resolve(UrlGenerator::class)->to('forum')->base()),
```
I believe we had issues in the past when resolving `UrlGenerator` early in Listener `__construct` methods. 

I can't find all the info regarding this issue, but I did find this commit from Franz in flarum/mentions (https://github.com/flarum/mentions/commit/4dc8068fc335d99b630ac0915cd2ffa67c696f4e) that stopped resolving UrlGenerator early in an event subscriber class.
Blomstra also came across this issue more recently, so it still exists: https://github.com/blomstra/flarum-ext-fontawesome/commit/14c55e5a3b0a6eff1e0229eaed3e90c84e40e71c.

This default will have to be set in another way that doesn't resolve UrlGenerator too early. I don't know the best way to do that right now that avoids complicating this simple code.

PatrickRoethlisberger replied:
I've had a chat with @imorland and he mentioned that for this to work, the `flarum\core` has to be `^1.3.1`. Could this be linked to the issue mentioned here? https://github.com/flarum/framework/pull/3439

imorland replied:
No, the 1.3.1 issue is different to this. @datitisev is referring to the early resolution of the `UrlGenerator` here.

The < 1.3.1 issue is chaining multiple `->default` settings were not registered correctly

dsevillamartin replied:
I don't think so. The UrlGenerator resolution problem here isn't related to the Settings extender itself, as you are calling `resolve` as a parameter to `->default()`. It doesn't matter how the extender handles that output as the resolution occurs in `extend.php` before the extender does anything.

I'm not entirely sure when the Sentry `init` function is called in the lifecycle, so I don't know if that would be after app boot. It might work there or it might need code in boot to run the Sentry `init` (I don't know if this can be run multiple times).

PatrickRoethlisberger replied:
Would you prefer if I move the UrlGenerator default to the `SentryJavaScript.php` and `SentryServiceProvider.php` files if no custom environment is set in the admin settings?

dsevillamartin replied:
Yep. Looks like that should work.

I added some logs to see the run order and SentryJavascript runs well after boot. And the Sentry `init` code only runs when Sentry is actually triggered (e.g. by an error, I had a `SentryServiceProvider@register HubInterface` log line that never appeared) so it should be fine.

```
[2022-07-28 13:29:43] flarum.DEBUG: SentryServiceProvider@register
[2022-07-28 13:29:43] flarum.DEBUG: SentryServiceProvider@boot
[2022-07-28 13:29:44] flarum.DEBUG: SentryJavascript
```

PatrickRoethlisberger replied:
Does this work for you?

PatrickRoethlisberger replied:
@datitisev just tell me when I can do something here

dsevillamartin replied:
@PatrickRoethlisberger I think this should work.

@imorland could you check as well whether the fix should solve the issues? 
I mainly checked load order but didn't test it with an ext that would be broken with the previous code.

PatrickRoethlisberger committed:
Fix issue with urlGenerator in extend.php

Moving urlGenerator into SentryServiceProvider and SentryJavaScript so default is not set in extend.php
